sources/osgShader/ShaderGenerator.js

```diff
@@ -203,13 +215,17 @@ define( [
 
             return function ( state ) {
                 // extract valid attributes
-                var hash = '';
+
+                this._hashStack.current = 0;
```

Comment:
It's bad practice to add attribute to an array.
Add an additional variable if needed.
Also... can the array be a global (global to all the ShaderGenerator)?

Anyway in that case I don't think it's super useful to avoid allocation...
We only push to this hash on a rendering leaf... (so typically a small number) and we only push a small of a stuffs (number of attributes).

I did some test a while ago by using a similar array workaround to avoid push/pop... and for relatively small array, I didn't notice any gain (probably because of the similar pushToHashStack and because js VM handles well small array).
But still, I'm against useless push/pop :).